### PR TITLE
Reduce the goroutine in flowgraph to 2

### PR DESCRIPTION
**Pull request by smellthemoon**
#24826 
/kind improvement

### CODE REVIEW

File: internal/datanode/data_sync_service.go [hunk 1]
```diff
@@ -81,12 +81,18 @@ type nodeConfig struct {
...
-// start the flow graph in datasyncservice
+// start the flow graph in dataSyncService
 func (dsService *dataSyncService) start() {
 	if dsService.fg != nil {
 		log.Info("dataSyncService starting flow graph", zap.Int64("collectionID", dsService.collectionID),
 			zap.String("vChanName", dsService.vchannelName))
-		dsService.fg.Start()
+		// won't be stuck here
+		var wg sync.WaitGroup
+		wg.Add(1)
```
Comment (yah01) [inline internal/datanode/data_sync_service.go:91]: You don't need wait group if never wait for the goroutine done

File: internal/util/flowgraph/node.go [hunk 2]
```diff
@@ -69,8 +69,14 @@ type nodeCtx struct {
 func (nodeCtx *nodeCtx) Start() {
 	nodeCtx.node.Start()
 
-	nodeCtx.closeWg.Add(1)
-	go nodeCtx.work()
+	// in dmInputNode, message from mq to channel, alloc goroutines
+	// limit the goroutines in other node to prevent huge goroutines numbers
+	if nodeCtx.node.IsInputNode() {
+		nodeCtx.closeWg.Add(1)
+		go nodeCtx.work()
```
Comment (yah01) [inline internal/util/flowgraph/node.go:76]: don't rely the callee to release the wait group, go this:
```golang
go func() {
    defer nodeCtx.closeWg.Done()
    nodeCtx.work()
}
```

**Review by yah01 — CHANGES_REQUESTED** [review]

File: internal/util/flowgraph/node.go [hunk 3]
```diff
@@ -53,26 +53,156 @@ type BaseNode struct {
 	maxParallelism int32
 }
 
+// manage nodeCtx
+type nodeCtxManager struct {
+	inputNodeCtx *nodeCtx
+	closeWg      *sync.WaitGroup
```
Comment (yah01) [inline internal/util/flowgraph/node.go:59]: passing `WaitGroup` into a struct is not a good idea, add 2 `StartXxx` methods and operates the `WaitGroup` in outside

File: internal/util/flowgraph/node.go [hunk 4]
```diff
@@ -53,26 +53,156 @@ type BaseNode struct {
...
+}
+
+func (nodeCtxManager *nodeCtxManager) inputNodeWork(checker *timerecord.GroupChecker) {
+	defer nodeCtxManager.closeWg.Done()
+	inputNode := nodeCtxManager.inputNodeCtx
+	for {
+		select {
+		case <-nodeCtxManager.inputNodeCloseCh:
+			// remove input node checker
+			checker.Remove(fmt.Sprintf("nodeCtxTtChecker-%s", inputNode.node.Name()))
```
Comment (yah01) [inline internal/util/flowgraph/node.go:94]: Also spelling the name is not a good idea, add a const string equals to the prefix, or remove the prefix

File: internal/util/flowgraph/node.go [hunk 5]
```diff
@@ -53,26 +53,156 @@ type BaseNode struct {
...
+	}
+}
+
+// Close handles cleanup logic and notify worker to quit
+func (nodeCtxManager *nodeCtxManager) Close() {
+	nodeCtx := nodeCtxManager.inputNodeCtx
+	for nodeCtx != nil {
+		nodeCtx.node.Close()
+		log.Debug("flow graph node closed", zap.String("nodeName", nodeCtx.node.Name()))
+		nodeCtx = nodeCtx.downstream
```
Comment (yah01) [inline internal/util/flowgraph/node.go:193]: this line does nothing, it assign a new address to the pointer `nodeCtx`, but not `nodeCtxManager.intputNodeCtx`. They are only two different pointer variables which point to the same address.

Also the closing way is not thread safe, and too tricky. Implementing the closing logic in node itself, the manager should only call the `Close()` of nodes, and wait it done

File: internal/util/flowgraph/node.go [hunk 6]
```diff
@@ -99,67 +229,14 @@ func isCloseMsg(msgs []Msg) bool {
...
-}
-
 // Close handles cleanup logic and notify worker to quit
 func (nodeCtx *nodeCtx) Close() {
 	if nodeCtx.node.IsInputNode() {
-		nodeCtx.node.Close()
+		for nodeCtx != nil {
+			nodeCtx.node.Close()
+			log.Debug("flow graph node closed", zap.String("nodeName", nodeCtx.node.Name()))
+			nodeCtx = nodeCtx.downstream
```
Comment (yah01) [inline internal/util/flowgraph/node.go:238]: Same as line 193

**Review by yah01 — CHANGES_REQUESTED** [review]

File: internal/util/flowgraph/node.go [hunk 7]
```diff
@@ -53,26 +53,156 @@ type BaseNode struct {
...
+		default:
+			// inputs from inputsMessages for Operate
+			var input, output []Msg
+			// inputNode.input not from nodeCtx.inputChannel
+			// the input message decides whether the operate method is executed
+			n := inputNode.node
+			inputNode.blockMutex.RLock()
+			if !n.IsValidInMsg(input) {
+				inputNode.blockMutex.RUnlock()
+				return
```
Comment (yah01) [inline internal/util/flowgraph/node.go:111]: just continue

File: internal/util/flowgraph/node.go [hunk 8]
```diff
@@ -53,26 +53,156 @@ type BaseNode struct {
...
+			for curNode != nil {
+				// inputs from inputsMessages for Operate
+				var input, output []Msg
+				input = <-curNode.inputChannel
+				// the input message decides whether the operate method is executed
+				n := curNode.node
+				curNode.blockMutex.RLock()
+				if !n.IsValidInMsg(input) {
+					curNode.blockMutex.RUnlock()
+					return
```
Comment (yah01) [inline internal/util/flowgraph/node.go:161]: same as line 111

File: internal/util/flowgraph/node.go [hunk 9]
```diff
@@ -53,26 +53,156 @@ type BaseNode struct {
...
+			// the output decide whether the node should be closed.
+			if isCloseMsg(output) {
+				close(nodeCtxManager.inputNodeCloseCh)
+				if inputNode.inputChannel != nil {
+					close(inputNode.inputChannel)
+				}
+				inputNode.Close()
+			}
+			// deliver to all following flow graph node.
+			inputNode.downstream.inputChannel <- output
```
Comment (yah01) [inline internal/util/flowgraph/node.go:124]: you need to change the inputChannel to a buffered channel, with size at least 1

**Review by yah01 — CHANGES_REQUESTED** [review]

File: internal/util/flowgraph/node.go (hunk 9 above)
Comment (smellthemoon) [reply internal/util/flowgraph/node.go:124]: this default size is 16

File: internal/util/flowgraph/node.go (hunk 3 above)
Comment (smellthemoon) [reply internal/util/flowgraph/node.go:59]: WaitGroup is fg.closeWg, fg.close will do this.

**Pull request closed without merging**